### PR TITLE
[Tests] Increase QMK test coverage

### PR DESCRIPTION
**Pull request by KarlK90**
## Description

This PR enhances the current ~~unit-testing~~ integration testing facilities and adds additional tests:

* All tests use an isolated per-test keymap instead of a global shared keymap in `keymap.c` the usage goes like this:

```cpp
/* Define keys like this: layer, column, row, keycode */
auto       key_a      = KeymapKey(0, 0, 0, KC_A);
auto       key_lshift = KeymapKey(0, 1, 1, KC_LSFT);

/* Add keys to keymap for this test */
set_keymap({key_a, key_lshift});

key_a.press();
key_a.release();
key_lshift.press();
key_lshift.release();
```

* Tests that fail provide a lot more context on the events that happened during the execution.
* Tests were added:
  * basic key-presses
  * mod-taps and layer taps with different tap-hold configurations
  * one-shot mods and layers

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by KarlK90** [comment]
This is now ready for review and a good starting point to add more test-cases.

I haven't figured out a good way to run the different configurations in one go as this requires a change to the way the tests are compiled and called by the make system. Maybe one of the reviewers has a killer idea? The way QMK currently works would require several differently configured testing binaries to be build and run. For instance one with the default mod-tap behavior, one with mod-tap-interrupt only, one with force-tap-hold etc.

**Comment by KarlK90** [comment]
Every QMK configuration that needs a different binary is now split out into an individual testing binary, so all tests can be run simultaneously. :muscle:

**Comment by drashna** [comment]
Can the "in progress" label be removed here?

**Comment by KarlK90** [comment]
Yes it can!

**Comment by drashna** [comment]
Looks like this is erroring out on: 
```
Unexpected mock function call - returning directly.
    Function call: send_keyboard_mock(@0x5582251ce890 Keyboard Report: Mods (0) Keys ()
)
Google Mock tried the following 2 expectations, but none matched:

tests/basic/test_action_layer.cpp:286: tried expectation #0: EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_RALT)))...
  Expected arg #0: is equal to Keyboard Report: Mods (64) Keys ()

           Actual: Keyboard Report: Mods (0) Keys ()

         Expected: to be called once
           Actual: never called - unsatisfied and active
tests/basic/test_action_layer.cpp:287: tried expectation #1: EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_RALT, KC_9)))...
```

Running this from the docker container.

**Comment by KarlK90** [comment]
Yes, this is expected, as the test documents false buggy behavior and was added in https://github.com/qmk/qmk_firmware/pull/13789/commits/d50dcafaffd4d991ebcce0d65d6218a1d5e10c12.

**Comment by drashna** [comment]
Well, specifically it is causing one of the tests to fail:
```
[  PASSED  ] 49 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] ActionLayer.LayerTapReleasedBeforeKeypressReleaseWithModifiers
```

**Comment by KarlK90** [comment]
Yup, that is the one I added in the commit. It is an edge-case which doesnt happen to often on ansi-us layouts and the like. My reasoning was that I rather document the expected behaviour then what QMK currently does. Which is removing the modifier from the report on a layer change, altough it should not be modified because the key adding the modifier is still pressed. This results on a german layout in ((((\<layer release\>8888 instead of ((((\<layer release\>((((

**Review by drashna — APPROVED** [review]

File: tests/tap_hold_configurations/default_mod_tap/test_tap_hold.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_A, KC_P)));
+    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_P)));
+    EXPECT_CALL(driver, send_keyboard_mock(_));
+    layer_tap_hold_key.release();
+    run_one_scan_loop();
+    testing::Mock::VerifyAndClearExpectations(&driver);
+}
+
+TEST_F(DefaultTapHold, tap_mod_tap_hold_key_two_times) {
+    GTEST_SKIP() << "Keymap out of bounds access, uncomment for diagnostics!";
```
Comment (tzarc) [inline tests/tap_hold_configurations/default_mod_tap/test_tap_hold.cpp:143]: Can you clarify, here? Testing bug? QMK bug?

File: tests/tap_hold_configurations/default_mod_tap/test_tap_hold.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+    /* Release mod-tap-hold key. */
+    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport()));
+    mod_tap_hold_key.release();
+    run_one_scan_loop();
+    testing::Mock::VerifyAndClearExpectations(&driver);
+}
+
+TEST_F(DefaultTapHold, tap_mod_tap_hold_key_twice_and_hold_on_second_time) {
+    GTEST_SKIP() << "Keymap out of bounds access, uncomment for diagnostics!";
```
Comment (tzarc) [inline tests/tap_hold_configurations/default_mod_tap/test_tap_hold.cpp:178]: Can you clarify, here? Testing bug? QMK bug?

File: tests/tap_hold_configurations/retro_tapping/test_tapping.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport()));
+    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_P)));
+    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport()));
+    mod_tap_hold_key.release();
+    run_one_scan_loop();
+    testing::Mock::VerifyAndClearExpectations(&driver);
+}
+
+TEST_F(Tapping, ANewTapWithinTappingTermIsBuggy) {
```
Comment (tzarc) [inline tests/tap_hold_configurations/retro_tapping/test_tapping.cpp:52]: Buggy how? Clarification would be great.

File: tmk_core/native.mk [hunk 4]
```diff
@@ -31,4 +37,4 @@ CFLAGS += -fno-strict-aliasing
 
 CXXFLAGS += $(COMPILEFLAGS)
 CXXFLAGS += -fno-exceptions
-CXXFLAGS += -std=gnu++11
+CXXFLAGS += -std=gnu++17
```
[the remark this answers is not in the document]
Comment (tzarc) [reply tmk_core/native.mk:34]: Pretty sure we ended up converging on C++14 given the reliance on Debian's bundling of gcc 5.4.0.
And yes, I agree with @zvecr -- if we do intend on bumping the version, it should be everywhere.

**Review by tzarc — CHANGES_REQUESTED** [review]

**Comment by KarlK90** [comment]
Thanks @tzarc and @zvecr for the feedback, I'll have more time next week during my holiday. Looking forward to getting this merged.

File: tmk_core/native.mk (hunk 4 above)
[the remark this answers is not in the document]
Comment (KarlK90) [reply tmk_core/native.mk:34]: Good point, I mainly bumped to C++17 to use `std::optional` but converted to code to use plain pointers instead. A C++ bump for all components should be in the scope of a different pr.

File: tests/tap_hold_configurations/retro_tapping/test_tapping.cpp (hunk 3 above)
Comment (KarlK90) [reply tests/tap_hold_configurations/retro_tapping/test_tapping.cpp:52]: I would have to dig into this test myself, as it was originally written by @fredizzimo

File: tests/basic/test_action_layer.cpp [hunk 5]
```diff
@@ -90,3 +91,275 @@ TEST_F(ActionLayer, LayerStateCmp) {
...
+    TestDriver driver;
+    KeymapKey  layer_key = KeymapKey{0, 0, 0, TT(1)};
+
+    set_keymap({layer_key});
+
+    /* Press and release TT. */
+    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport())).Times(0);
+    layer_key.press();
+    run_one_scan_loop();
+    EXPECT_TRUE(layer_state_is(0)); /* This is wrong? */
```
[the remark this answers is not in the document]
Comment (KarlK90) [reply tests/basic/test_action_layer.cpp:221]: Exactly but it is actually 0 which should be false, very curios. I'll add the correct expected behavior but set the test to be ignored and collect those in a separate issue.

**Comment by drashna** [comment]
Some merge conflicts

**Review by tzarc — APPROVED** [review]

**Comment by tzarc** [comment]
As discussed on Discord, reverted thru commit 7746aefe94cc4cd492bfafdef73d95d073f0603b.
Follow-up PR to come.
